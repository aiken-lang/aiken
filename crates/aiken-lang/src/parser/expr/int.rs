use chumsky::prelude::*;

use crate::{
    expr::UntypedExpr,
    parser::{error::ParseError, literal::uint, token::Token},
};

pub fn parser() -> impl Parser<Token, UntypedExpr, Error = ParseError> {
    uint().map_with_span(|(value, base), span| UntypedExpr::UInt {
        location: span,
        value,
        base,
    })
}

#[cfg(test)]
mod tests {
    use crate::assert_expr;

    #[test]
    fn int_literal() {
        assert_expr!("1");
    }

    #[test]
    fn int_negative() {
        assert_expr!("-1");
    }

    #[test]
    fn int_numeric_underscore() {
        assert_expr!(
            r#"
            {
              let i = 1_234_567
              let j = 1_000_000
              let k = -10_000
            }
            "#
        );
    }

    #[test]
    fn int_hex_bytes() {
        assert_expr!(r#"0x01"#);
    }
}
