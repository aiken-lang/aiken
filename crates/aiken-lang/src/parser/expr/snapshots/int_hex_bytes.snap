---
source: crates/aiken-lang/src/parser/expr/int.rs
description: "Code:\n\n0x01"
---
UInt {
    location: 0..4,
    value: "1",
    base: Hexadecimal,
}
